File: Find Min (Stack)/ruby-solution/min_stack.rb
require_relative 'node'

class MinStack

  attr_accessor :top

  def initialize
    @top = nil
  end

  def pop
    if @top
      item = @top.data
      @top = @top.next
      return item
    else
      puts "Empty Stack"
    end
  end

  def push(data)
    new_top = nil
    if !@top || data < @top.data
      new_top = Node.new(data)
    else
      new_top = Node.new(@top.data)
    end
    new_top.next = @top
    @top = new_top
  end

  def peek
    if top
      @top.data
    else
      puts "Empty Stack"
    end
  end

end

File: Find Min (Stack)/python-solution/minstack.py
from node import Node

class MinStack(object):
  def __init__(self):
    self.top = None

  def pop(self):
    if self.top == None:
      raise SystemExit("Empty Stack")
    item = self.top.data
    self.top = self.top.next
    return item

  def push(self, data):
    new_top = None
    if self.top == None or data < self.top.data:
      new_top = Node(data)
    else:
      new_top = Node(self.top.data)
    new_top.next = self.top
    self.top = new_top

  def peek(self):
    if self.top == None:
      raise SystemExit("Empty Stack - MinStack Empty, can't peek")
    else:
      return self.top.data